File: .github/scripts/fetch_news.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

const (
	newsAPIURL = "https://newsapi.org/v2/top-headlines?country=mx&pageSize=1"
	contentDir = "content"
)

type NewsResponse struct {
	Status       string `json:"status"`
	TotalResults int    `json:"totalResults"`
	Articles     []struct {
		Title       string `json:"title"`
		Description string `json:"description"`
		Content     string `json:"content"`
		URL         string `json:"url"`
		Source      struct {
			Name string `json:"name"`
		} `json:"source"`
		PublishedAt string `json:"publishedAt"`
	} `json:"articles"`
}

type ErrorResponse struct {
	Status  string `json:"status"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

func main() {
	apiKey := os.Getenv("NEWSAPI_KEY")
	if apiKey == "" {
		panic("NEWSAPI_KEY no está definido")
	}

	// Construir URL con API key
	url := fmt.Sprintf("%s&apiKey=%s", newsAPIURL, apiKey)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		panic(fmt.Sprintf("Error creando request: %v", err))
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		panic(fmt.Sprintf("Error haciendo request: %v", err))
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		panic(fmt.Sprintf("Error leyendo respuesta: %v", err))
	}

	// Intentar decodificar como error primero
	var errorResp ErrorResponse
	if err := json.Unmarshal(body, &errorResp); err == nil && errorResp.Status == "error" {
		panic(fmt.Sprintf("Error de API: %s - %s", errorResp.Code, errorResp.Message))
	}

	// Si no es error, decodificar como respuesta normal
	var news NewsResponse
	if err := json.Unmarshal(body, &news); err != nil {
		panic(fmt.Sprintf("Error decodificando respuesta: %v", err))
	}

	if news.Status != "ok" {
		panic(fmt.Sprintf("Estado de respuesta no OK: %s", news.Status))
	}

	if len(news.Articles) == 0 {
		fmt.Println("No se encontraron noticias.")
		return
	}

	article := news.Articles[0]
	date := time.Now().Format("2006-01-02")
	filename := fmt.Sprintf("%s/news-%s.md", contentDir, date)

	mdContent := fmt.Sprintf(`# %s

**Fuente:** [%s](%s)  
**Fecha:** %s

%s

%s
`, article.Title, article.Source.Name, article.URL, article.PublishedAt[:10], article.Description, article.Content)

	if err := ioutil.WriteFile(filename, []byte(mdContent), 0644); err != nil {
		panic(fmt.Sprintf("Error guardando archivo: %v", err))
	}

	fmt.Println("Noticia guardada en:", filename)
}

File: .github/scripts/fetch_news.py
import os
import requests
from datetime import datetime
from pathlib import Path

NEWS_API_KEY = os.getenv("NEWSAPI_KEY")
CONTENT_DIR = "content"
QUERY = "noticias -deportes"
LANGUAGE = "es"
PAGE_SIZE = 5

def fetch_news():
    if not NEWS_API_KEY:
        raise Exception("Falta la variable de entorno NEWSAPI_KEY")

    url = "https://newsapi.org/v2/everything"
    params = {
        "q": QUERY,
        "language": LANGUAGE,
        "sortBy": "publishedAt",
        "pageSize": PAGE_SIZE,
        "apiKey": NEWS_API_KEY
    }

    response = requests.get(url, params=params)
    data = response.json()

    if data.get("status") != "ok":
        raise Exception(f"Error de la API: {data.get('code')} - {data.get('message')}")

    articles = data.get("articles", [])
    if not articles:
        print("🟡 No se encontraron artículos.")
        return

    Path(CONTENT_DIR).mkdir(parents=True, exist_ok=True)
    date_str = datetime.now().strftime("%Y-%m-%d")

    for i, article in enumerate(articles):
        filename = f"{CONTENT_DIR}/news-{date_str}-{i+1}.md"
        content = f"""# {article['title']}

**Fuente:** [{article['source']['name']}]({article['url']})  
**Fecha:** {article['publishedAt'][:10]}

{article.get('description', '')}

{article.get('content', '')}
"""
        with open(filename, "w", encoding="utf-8") as f:
            f.write(content)

        print(f"✅ Noticia guardada en: {filename}")

if __name__ == "__main__":
    fetch_news()
